util/oci/client.go

```diff
@@ -0,0 +1,302 @@
+	return func(c *nativeOciClient) {
+		c.indexCache = indexCache
+	}
+}
+
+func WithChartPaths(repoCachePaths argoio.TempPaths) ClientOpts {
+	return func(c *nativeOciClient) {
+		c.repoCachePaths = repoCachePaths
+	}
+}
+func NewClient(repoURL string, creds Creds, proxy string, opts ...ClientOpts) (Client, error) {
+	return NewClientWithLock(repoURL, creds, globalLock, proxy, opts...)
+}
+func NewClientWithLock(repoURL string, creds Creds, repoLock sync.KeyLock, proxyUrl string, opts ...ClientOpts) (Client, error) {
+	ociRepo := strings.TrimPrefix(repoURL, "oci://")
+	repo, err := remote.NewRepository(ociRepo)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize repository: %v", err)
+	}
+
+	repo.PlainHTTP = creds.InsecureHttpOnly
+
+	ociUri, err := url.Parse(repoURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse registry host: %v", err)
+	}
+
+	var tlsConf *tls.Config
+	if !repo.PlainHTTP {
+		tlsConf, err = newTLSConfig(creds)
+		if err != nil {
+			return nil, fmt.Errorf("failed setup tlsConfig: %v", err)
+		}
+	}
+
+	var ociRegistry string
+	if ociUri.Port() != "" {
+		ociRegistry = fmt.Sprintf("%s:%s", ociUri.Hostname(), ociUri.Port())
+	} else {
+		ociRegistry = ociUri.Hostname()
+	}
+
+	client := &http.Client{Transport: &http.Transport{
+		Proxy:             proxy.GetCallback(proxyUrl),
+		TLSClientConfig:   tlsConf,
+		DisableKeepAlives: true,
+	}}
+	repo.Client = &auth.Client{
+		Client: client,
+		Cache:  nil,
+		Credential: auth.StaticCredential(ociRegistry, auth.Credential{
+			Username: creds.Username,
+			Password: creds.Password,
+		}),
+	}
+
+	c := &nativeOciClient{
+		creds:    creds,
+		repoURL:  ociRepo,
+		proxy:    proxyUrl,
+		repoLock: repoLock,
+		repo:     repo,
+	}
+	for i := range opts {
+		opts[i](c)
+	}
+	return c, nil
+}
+
+// nativeOciClient implements Client interface using oras-go
+type nativeOciClient struct {
+	creds          Creds
+	repoURL        string
+	proxy          string
+	repo           *remote.Repository
+	repoLock       sync.KeyLock
+	indexCache     indexCache
+	repoCachePaths argoio.TempPaths
+}
+
+func (c *nativeOciClient) TestRepo(ctx context.Context) (bool, error) {
+	err := c.repo.Tags(ctx, "", func(tags []string) error {
+		return nil
+	})
+	return err == nil, err
+}
+
+func (c *nativeOciClient) Extract(ctx context.Context, revision string, project string, manifestMaxExtractedSize int64, disableManifestMaxExtractedSize bool) (string, argoio.Closer, error) {
+	cachedPath, err := c.getCachedPath(revision, project)
+	if err != nil {
+		return "", nil, err
+	}
+
+	c.repoLock.Lock(cachedPath)
+	defer c.repoLock.Unlock(cachedPath)
+
+	exists, err := fileExist(cachedPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	fs, err := file.New(cachedPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if !exists {
+		_, err = oras.Copy(ctx, c.repo, revision, fs, revision, oras.DefaultCopyOptions)
+		if err != nil {
+			return "", nil, err
+		}
+	}
```

Comment:
Do we have a better way like a checkout functionality as part of the `oras-go` than manually performing the file exists check here? If not, does it make sense to add it?

Also, maybe should we use https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/registry/remote#Repository.Fetch instead of `Copy`?

@blakepettersson @Wwwsylvia

Reply:
Dumb question: is `cachedPath` expected be a path to a file or a directory?

The parameter of [`file.New()`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/content/file#New) is supposed to be directory.
When the artifact is copied to `fs`, the layers with the `org.opencontainers.image.title` annotation will be created in the file system, while layers without this annotation will be preserved in the memory. If the file already exists in the file system it will be overwritten, unless `DisableOverwrite` is set.

Reply:
> Dumb question: is `cachedPath` expected be a path to a file or a directory?

Originally I intended this to be a directory, but my current thinking is that this should now be a tar file of the OCI store contents. The reason for this is that due to the directory permissions of its parent, `os.RemoveAll` doesn't actually work when removing a directory in the repo-server cache dir (but works fine when deleting a file :shrug:). 

My current idea is basically this:

1. On the first download, download the content and save it as a tar file in `cachedPath`. This file would not be touched again unless the repository cache gets hard refreshed (a hard refresh will delete the tar file so it gets downloaded again from the remote store). 
2. Afterwards (and for all subsequent extractions of a given digest), initialize a new OCI store using `oci.NewFromTar` from `cachedPath`.
3. Extract the manifest content from the local OCI store to a temp dir. This is what the repo-server would use to generate the manifests, which will ultimately be cached until the next refresh (by default 12h or if a user initiates a hard refresh).

> Repository.Fetch is for getting the manifest content 

I only want the content, so that suits me just fine.

Reply:
Sorry I don't have much context on `Extract`, what's the purpose of it? Is it just for caching?

> My current idea is basically this:
> 1. On the first download, download the OCI content and save it as a tar file in cachedPath. This file would not be touched again unless the repository cache gets hard refreshed (a hard refresh will delete the tar file so it gets downloaded again from the remote store).
> 2. Afterwards (and for all subsequent extractions of a given digest), initialize a new OCI store using oci.NewFromTar from cachedPath.
> 3. Extract the manifest content from the local OCI store to a temp dir. This is what the repo-server would use to generate the manifests, which will ultimately be cached until the next refresh (by default 12h or until a user initiates a hard refresh).

For step 3, what does it mean by "extract the manifest content to a temp dir"? In what format would the manifest be present in the temp dir? Like what would be the file name? How about the config and the layers? And for "which will ultimately be cached", is the cache refers to server cache?

Reply:
It sounds like we want to do something like this?

```go
// 1. On the first download, download the OCI content and save it as a tar file
ociDir, err := oci.New(cachedPath)
tarball, err := tarDirectory(cachedPath)

// 2. Afterwards (and for all subsequent extractions of a given digest), initialize a new OCI store using oci.NewFromTar from cachedPath.
readonlyOCI, err = oci.NewFromTar(cachedPath)

// 3. Extract the manifest content from the local OCI store to a temp dir. 
tempDir := getTempDir()
fs, err := file.New(tempDir)
// copy the whole artifact
desc, err := oras.Copy(ctx, readonlyOCI, "mytag", fs, "mytag", oras.DefaultCopyOptions)
```

Reply:
> I would like to get the files which are meant to be used for syncing to a cluster. I can get those easily with oras pull. This would mean I would need to use Copy right? Otherwise, how does it work with Repository.Fetch?

Yes. `oras pull` is equivalent to 
```go
fs, err := file.New(workingDir)
desc, err := oras.Copy(ctx, repo, "mytag", fs, "mytag", oras.DefaultCopyOptions)
```
and `oras pull --oci-layout` is equivalent to

```go
ociStore, err := oci.New(workingDir)
desc, err := oras.Copy(ctx, repo, "mytag", ociStore, "mytag", oras.DefaultCopyOptions)
```

`Repository.Fetch` is for fetching a manifest/blob itself, it can be used if we want to print out the manifest content, etc.

Reply:
>It sounds like we want to do something like this?

I believe that's exactly what I'm looking for - I just added a new commit which pretty much does just that.

Reply:
> Sorry I don't have much context on Extract, what's the purpose of it? Is it just for caching?

It's a bit confusing what goes on in the repo-server, but the gist of what happens when an app is refreshed (this happens every 3 minutes by default) is this:

1. The repo-server will attempt resolve an abstract ref to a SHA/digest.
2. It will then see if for the given SHA/digest if there are already pre-cached manifests available for said SHA/digest. This is coming from Redis, and is valid for 12h by default. If available, short-circuit and return cached manifests to the caller.
3. If there is nothing in the cache, either due to the cache having expired or to the given digest having never been downloaded in the first place, it will check to see if there's a local Helm chart/Git repo/OCI image on local disk. This is stored in a random location, which can only be looked up with the correct combo of `repoUrl` + `appProject` + `digest`.
4. If it's not present, download the relevant Helm chart/Git repo/OCI image.
5. Afterwards, extract the contents of the artifact to a randomised path on disk
6. Then the repo-server will generate the manifests from the extracted randomised path, where applicable (e.g `helm template`, `kustomize build` etc etc).
7. The results of _that_ are then stored in Redis and returned to the caller.

This is quite simplified, and I'm not sure I got it 100% correct but it's more or less the case. So, to answer your question, the responsibility of `Extract` in the context of OCI would be to handle points 3-5.

Reply:
@blakepettersson Thanks for the clarification! Some follow-up questions:

For point 3,
> This is stored in a random location, which can only be looked up with the correct combo of repoUrl + appProject + digest.
Is `cachedPath` the location (repoUrl + appPojrect + digest)?

For point 6, 
> Then the repo-server will generate the manifests from the extracted randomised path

Is the "manifest" an OCI manifest of the artifact or something else? I guess it is something else? As we already have the manifest of artifact since step 4.

Reply:
@Wwwsylvia

Reply:
Got it. Thanks a lot!